TripalCultivate/TripalCultivate-Phenotypes
# Updates file template generator service

**Issue #124** Automated test needed for describeUploadFileFormat().

## Motivation
This PR updates the template generator service and automated test to accommodate customization of file extension of the generated template file. The previous version supports a single file extension and is hard-coded into the logic of the class. In addition, an unused variable $notes in https://github.com/TripalCultivate/TripalCultivate-Phenotypes/blob/3d934ca9215e046bd890e00cce6f2388cd19445f/trpcultivate_phenotypes/src/Plugin/TripalImporter/TripalCultivatePhenotypesTraitsImporter.php#L1488 was not implemented but is now added into the header information section (see image below).

<img width="944" alt="image" src="https://github.com/user-attachments/assets/2e8821ad-282e-41db-b806-3da47cd23cd0" />




## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an automated test to ensure it doesn't return.*

1. Display additional notes in the headers information section of the Importer form.
2. Replace the static file extension by improving the logic of determining the supported file types.
3. Reference static variables in validator base and validator file types to establish MIME type and delimiter.
4. Upgrade automated test.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

ClassName->testTemplateGeneratorService()
- file_types annotation definition with single item - tsv.
- file_types annotation definition with multiple items but only the first item is preferred.
- A file type (txt - text/plain) where the mapping of delimiter contains multiple characters but only the first item is selected.

For every test case, the template generator is check to:
1. It creates the file uri to the template file.
2. The template file is stored in the configured directory for template files.
3. Filename is a combination of the plugin id, a string, current user username.
4. Filename contains the plugin id.
5. Filename contains the username.
6. File extension is the expected file extension.
7. The header row matches the headers array if it was encoded as string using the delimiter.
8. Separating the header row string should match the raw headers array.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Switch to g2.124-describeUploadFileFormat
2. In a fully setup site, paste the code below in your site/devel/php
```
$template_generator = \Drupal::service('trpcultivate_phenotypes.template_generator');

$column_headers = [
  'Header A',
  'Header B',
  'Header C'
];

$importer_id = 'my-importer';

$file_extensions = [
  'tsv',
  'csv'
];

$l = $template_generator->generateFile($importer_id, $column_headers, $file_extensions);

dpm($l);

```
3. Copy the link and paste in the location bar in another browser window.
<img width="723" alt="image" src="https://github.com/user-attachments/assets/ed4256e2-2c2d-48ff-a56a-0801db5d3a64" />
4. Append the uri with the host.
localhost/site/default/file....
5. Click enter and save file when prompted. Adjust parameters to test.
